pkg/locker/locker.go

```diff
@@ -90,6 +93,13 @@ func (l *Locker) Lock(name string) {
 	// once locked then we can decrement the number of waiters for this lock
 	nameLock.Lock()
```

Comment:
Yes. All goroutines block here, but only one can do `CancelWithError`, this goroutine will unlock the lock then. After that, one of the other goroutines will continue running from this line, and it checks the error and returns the error. Though the waiters are canceled one by one, it is very quick because the judgement logic is inside of locker.Locker("name"), just like the lock is canceled.

Another way, like you said, I've thought about using `sync.Cond` to replace the `sync.Mutex`. 

```go
locker.Lock("name")
if locker.IsFirstOneGetTheLock("name") {   // Maybe mix sync.Once to do this
  // do the stuff, get the plugin (wait 16s), get err
  if err != nil {
    locker.SetError("name", err)
  }
  locker.Broadcast("name")
} else {
  for condition()  {    // How to define the condition here?
    locker.Wait("name")
  }
  locker.Unlock("name)
}
```

Its advantage is that `pkg/locker` looks much more like a "Lock" package. However, it is more complicated and harder to be used.